src/libpriv/rpmostree-util.c

```diff
@@ -226,29 +226,112 @@ rpmostree_translate_path_for_ostree (const char *path)
   return NULL;
 }
 
+/* Check that a date format is one of the supported directives.
+ * Supported directives must all be numeric and with a definite
+ * number of digits as defined at http://www.cplusplus.com/reference/ctime/strftime/.
+ * Note: year Y is treated as fixed size. */
+static bool
+is_valid_date_field(const char *date_fmt)
+{
+  if (!date_fmt)
+    return FALSE;
+  const char *supported_directives = "CdgGHIjmMSUVwWyY";
+  for (int i = 0; i < strlen(date_fmt); i++)
+    {
+      if (i % 2 == 0 && i != (strlen(date_fmt) - 1) && date_fmt[i] == '%')
+        continue;
+      else if (!strchr(supported_directives, date_fmt[i]))
+        return FALSE;
+    }
+  return TRUE;
+}
+
+#define VERSION_TAG_REGEX "<(\\b[a-zA-Z]+\\b):\\s*(.*)?>"
+#define DATE_BUFFER_SIZE 32
+
+/* Get the next version, given a version prefix and a last version.
+ * auto_version_prefix may contain a `date` field, to indicate the
+ * current date to appear in the next version. Appends a version number
+ * to the version string returned, if it is not present in last_version;
+ * if it is present, the version number is incremented.
+ * 
+ * Returns the next version string if successful, and NULL in case of error.
+ */
 char *
-_rpmostree_util_next_version (const char *auto_version_prefix,
-                              const char *last_version)
+_rpmostree_util_next_version (const char   *auto_version_prefix,
+                              const char   *last_version,
+                              GError      **error)
 {
+  g_autoptr(GRegex) tag_regex = NULL;
+  g_autoptr(GMatchInfo) tag_match_info = NULL;
+  g_autofree char *tag = NULL;
+  g_autofree char *value = NULL;
+  const char *new_prefix = NULL;
   unsigned long long num = 0;
   const char *end = NULL;
 
-  if (!last_version || !g_str_has_prefix (last_version, auto_version_prefix))
-    return g_strdup (auto_version_prefix);
+  tag_regex = g_regex_new (VERSION_TAG_REGEX, 0, 0, error);
+  if (!tag_regex)
+    return NULL;
 
-  if (g_str_equal (last_version, auto_version_prefix))
-    return g_strdup_printf ("%s.1", auto_version_prefix);
+  const bool match = g_regex_match (tag_regex, auto_version_prefix,
+                                    0, &tag_match_info);
+  bool date_given = FALSE;
+  if (match)
+    {
+      tag = g_match_info_fetch (tag_match_info, 1);
+      value = g_match_info_fetch (tag_match_info, 2);
+      g_assert (tag);
+      g_assert (value);
+      date_given = g_strcmp0 (tag, "date") == 0;
+    }
+  
+  if (date_given)
+    {
+      char date_buffer[DATE_BUFFER_SIZE] = {'\0'};
+      GTimeVal current_datetime;
+      GDate current_date;
+      g_get_current_time (&current_datetime);
+      g_date_set_time_val (&current_date, &current_datetime);
+      if (g_date_strftime (date_buffer, sizeof(date_buffer),
+                           value, &current_date) == 0
```

Comment:
Note that `0` just means the buffer was too small. Hmm, it's kinda weird that there isn't an API to return an allocated string, esp. since that function is already doing the `malloc` loop in there. Anyway, maybe let's just error out with "date string too long" for now?

Reply:
I initially saw that the function will return if the buffer is too small, but also noticed `0` was returned when `value == ""` - so I lumped both that case and `"buffer too small"` in with the message "`invalid date field"`. I'm good with erroring out with `"date string too long"` for now though.